packages/json-rpc-engine/src/JsonRpcEngine.ts

```diff
@@ -403,25 +414,23 @@ export class JsonRpcEngine extends SafeEventEmitter {
       return callback(error, {
         // Typecast: This could be a notification, but we want to access the
         // `id` even if it doesn't exist.
-        id: (callerReq as JsonRpcRequest).id ?? null,
+        id: (callerReq as JsonRpcRequest).id,
         jsonrpc: '2.0',
         error,
       });
-    }
-
-    // Handle notifications.
-    // We can't use isJsonRpcNotification here because that narrows callerReq to
-    // "never" after the if clause for unknown reasons.
-    if (this.#notificationHandler && !isJsonRpcRequest(callerReq)) {
+    } else if (
+      this.#notificationHandler &&
+      isJsonRpcNotification(callerReq) &&
+      !isJsonRpcRequest(callerReq)
```

Comment:
`isJsonRpcNotification` alone doesn't narrow `callerReq` to `JsonRpcNotification`.